raft rapid group recreation test & fix

tbg opened the pull request:
this simulates a rare error surfaced by #906 (and apparently others).
the test has a handful of concurrent actors add, commit to
(with retries) and remove a group.

currently this sometimes panics Raft:

E0506 13:09:45.552958   78550 log.go:188] applied(304) is out of range [prevApplied(10), committed(303)]
panic: applied(304) is out of range [prevApplied(10), committed(303)]

and other times (especially with a high number of writers) there are
deadlock issues.

i will continue to investigate these phenomena and hopefully attach
a fix to this PR prior to considering a merge.

tbg committed:
raft rapid group recreation test

this simulates a rare error surfaced by #906 (and apparently others).
the test has a handful of concurrent actors add, commit to
(with retries) and remove a group.

currently this sometimes panics Raft:

E0506 13:09:45.552958   78550 log.go:188] applied(304) is out of range [prevApplied(10), committed(303)]
panic: applied(304) is out of range [prevApplied(10), committed(303)]

and other times (especially with a high number of writers) there are
deadlock issues.

i will continue to investigate these phenomena and hopefully attach
a fix to this PR once it merges.

tschottdorf commented:
I think I have the culprit, and it's the asynchronous nature of writeTask in the MultiRaft loop. The following happens:

- Raft spits out a new Ready object. It's looked at it before, so it knows it's trying to have indexes 30-35 committed.
- a new WriteTask is started, it would bump the committed index from 30 to 35 but it hasn't done so yet.
- the group is deleted
- the group is recreated, but now it doesn't know about this pending WriteTask any more.
- the WriteTask finishes and the first thing the Raft group finds out is that someone is trying to move its applied index wayy past the committed index.

tschottdorf commented:
I've added a `writing` flag to each group that makes sure that a group only gets advanced by a finished writeTask if it also started that writeTask. That's now enough, however. A group that's just been deleted could still have put a `Ready` into the readyChan. This wouldn't be an issue with an unbuffered write, but the while the `multiNode.Ready()` channel is unbuffered, we sort of simulate a buffer because we read into `readyGroups` first - in effect a `Ready` from a previous incarnation of the group can get persisted and `Advance()`d onto the next (which hasn't read the updated stated from disk yet and crashes). I think we might have to add an internal "incarnation counter" to the groups, so this issue is really a Raft-internal incarnation of the stuff @bdarnell worries about in #756. Alternatively we need to change the loop so that a group cannot be deleted and re-added between consecutive `Ready` and `Advance()` stages.

tschottdorf commented:
I take that back, together with the write task change I've already made it should be enough if I just make sure to clean up the readyGroups when deleting a group. Will update PR shortly.

tschottdorf commented:
got it reliable now without dramatic changes. Will update the PR after lunch.

tbg committed:
fix raft races

- an outstanding writetask was able to be fed to a later incarnation of a group
- same for a Ready object

fixes #906 (at least in countless local interations - let's see what CircleCI
comes up with)

tbg committed:
avoid time.Sleep at end of test

seems not to be necessary any more since the race is fixed.
potentially CircleCI will still trip, though.

tschottdorf commented:
PTAL

bdarnell commented:
LGTM. Nice work tracking this down.
